storm-server/src/main/java/org/apache/storm/localizer/LocallyCachedBlob.java

```
@@ -0,0 +1,220 @@
+     * @param cb an optional callback indicating that they want to know/synchronize when a blob is updated.
+     */
+    public void addReference(final PortAndAssignment pna, BlobChangingCallback cb) {
+        if (cb == null) {
+            cb = NOOP_CB;
+        }
+        if (references.put(pna, cb) != null) {
+            LOG.warn("{} already has a reservation for {}", pna, blobDescription);
+        }
+    }
+
+    /**
+     * Removes a reservation for this blob from a given slot and assignemnt.
+     * @param pna the slot + assignment that no longer needs this blob.
+     */
+    public void removeReference(final PortAndAssignment pna) {
+        if (references.remove(pna) == null) {
+            LOG.warn("{} had no reservation for {}", pna, blobDescription);
+        }
+    }
+
+    /**
+     * Inform all of the callbacks that a change is going to happen and then wait for
+     * them to all get back that it is OK to make that change.
+     */
+    public synchronized void informAllOfChangeAndWaitForConsensus() {
+        CountDownLatch cdl = new CountDownLatch(references.size());
+        doneUpdating = new CompletableFuture<>();
+        for (Map.Entry<PortAndAssignment, BlobChangingCallback> entry : references.entrySet()) {
+            GoodToGo gtg = new GoodToGo(cdl, doneUpdating);
+            try {
+                PortAndAssignment pna = entry.getKey();
+                BlobChangingCallback cb = entry.getValue();
+                //TODO we probably want to not use this, or make it just return something that has less power to modify things
+                cb.blobChanging(pna.getAssignment(), pna.getPort(), this, gtg);
+            } finally {
+                gtg.countDownIfLatchWasNotGotten();
+            }
+        }
+        try {
+            cdl.await(3, TimeUnit.MINUTES);
+        } catch (InterruptedException e) {
+            //TODO need to think about error handling here in general.
```

Should we at least log the error being ignored?